### ISSUE TITLE
`nothrow` analysis not using all the information from type inference

### ISSUE THREAD

**Issue by oscardssmith** [JuliaLang/julia#47835]
For example, consider `Base.infer_effects(>>, (UInt32, Int32))` after #47830 or before #47567. Inference is able to figure out that the error path won't be hit, but the presence of the error path still poisons the effect analysis.

**Comment by aviatesk** [comment]
If inference is able to figure out that the error path won't be hit, then the `:nothrow` should not be tainted. Please let me know where exactly the `:nothrow` effect is tainted.

**Comment by oscardssmith** [comment]
I'm 90% sure the `:nothrow` is tainted on https://github.com/JuliaLang/julia/blob/c8662b593a245e3d1efa5b0d2b60175cfc23ebc7/base/operators.jl#L690

**Comment by aviatesk** [comment]
I don't think so -- can you check it with `Cthulhu.descend(...; with_effects=true)`?

**Comment by jakobnissen** [comment]
I've poked around a little. Julia's compiler can't prove that `>>(::UInt32, ::Int32)`  does not hit this line:
https://github.com/JuliaLang/julia/blob/c8662b593a245e3d1efa5b0d2b60175cfc23ebc7/base/operators.jl#L695

(This will error, as `-1` is not representable in an `UInt32`). Julia emits LLVM code including this error:
```
julia> code_llvm(>>, (UInt32, Int32); optimize=false)
[ lots of elided output]
L30:                                              ; preds = %L26
;  @ operators.jl:707 within `>>`
  %43 = call i32 @j_oftype_618(i32 zeroext %0, i64 signext -1) #0
  call void @llvm.trap()
  unreachable
```
But LLVM optimises it away. Hence, Julia is not able to figure out that it's nothrow.

**Comment by aviatesk** [comment]
Thanks so much @jakobnissen , very helpful. I think it's feasible to improve this analysis on Julia level.
